### PR TITLE
Module:snppileup

### PR DESCRIPTION
**Pull request by nikhil**
Includes module and tests for snppileup

### CODE REVIEW

**Comment by nikhil** [comment]
It looks like nf-test did not test my module, I need to figure out why

**Comment by nikhil** [comment]
I set branching rules so that only PRs that have been reviewed with tests passing can be merged to develop or main

**Comment by nikhil** [comment]
tests are now passing, this PR is ready for review. Related to #16

File: modules/msk/snppileup/main.nf
Comment (anoronh4) [inline modules/msk/snppileup/main.nf]: seems like it might be useful to join the tumor and normal into the same tuple. i think the way you have it written, when processing multiple tumor and normal samples, all elements in the channel will be paired with all other elements in the channel. so if you have 5 tumors and 5 normals, you will have 25 tasks launched, which might not be ideal.

**Review by anoronh4 — CHANGES_REQUESTED** [review]

**Comment by anoronh4** [comment]
could we also add a stub test?

File: modules/msk/snppileup/main.nf
Comment (buehlere) [inline modules/msk/snppileup/main.nf:37]: any way to dynamically grab these versions from the container or maybe not worth the trouble?

File: modules/msk/snppileup/main.nf
Comment (buehlere) [inline modules/msk/snppileup/main.nf]: do any optional args need to be specified positionally after the required file arguments? If so is adding an `args2` necessary?

**Review by buehlere — COMMENTED** [review]
I added a few small comments, but feel free to ignore. Once this comment is addressed: https://github.com/mskcc-omics-workflows/modules/pull/24/files#r1456026585, I think things look good.

**Comment by nikhil** [comment]
Do not merge yet, waiting on https://github.com/mskcc-omics-workflows/test-datasets/pull/2 to be merged. Once that is in, the test data files need to be updated. In the meantime, please verify all above issues have been met.

File: modules/msk/snppileup/main.nf
Comment (nikhil) [reply modules/msk/snppileup/main.nf]: Input channels have been updated, let me know if that works!

**Comment by nikhil** [comment]
> could we also add a stub test?

This should already be in: https://github.com/mskcc-omics-workflows/modules/blob/feature/snppileup/modules/msk/snppileup/main.nf#L42-L55

File: modules/msk/snppileup/main.nf
Comment (nikhil) [reply modules/msk/snppileup/main.nf:37]: Added the dynamic version generation

File: modules/msk/snppileup/main.nf
Comment (nikhil) [reply modules/msk/snppileup/main.nf]: There is no optional args after file arguments, so there is no need for an args2

**Comment by anoronh4** [comment]
> > could we also add a stub test?
> 
> This should already be in: https://github.com/mskcc-omics-workflows/modules/blob/feature/snppileup/modules/msk/snppileup/main.nf#L42-L55

I see a stub block in the process definition, but not a stub test in the main.nf.test file. i don't think it's a hard requirement, but it would be nice to have a check that the stub block works as expected. i'm going to approve and leave it to you to decide.

File: modules/msk/snppileup/main.nf
Comment (anoronh4) [reply modules/msk/snppileup/main.nf]: looks good to me!

**Comment by nikhil** [comment]
I added the stub test and fixed the test data, this PR is good to review now!

File: tests/config/test_data.config
Comment (buehlere) [inline tests/config/test_data.config]: Should we maybe write this in the style of nf-core?

Something roughly like this: 
```
test_data_base_msk = 'https://github.com/mskcc-omics-workflows/test-datasets/raw/'

test_data_msk {
 test_argos_sample_1_tumor_bam = "${params.test_data_base_msk}/bam_files/bams/argos_processed/sample_1_tumor.chr17.rg.md.abra.printreads.bam"

}
```
This might help keep things organized? 

Also there are a few quote and spacing edits in this file that I'm not sure we should merge? I guess they might be from the lint-ing though?

**Review by buehlere — CHANGES_REQUESTED** [review]
Looks good. I think we might want to update the formatting in the `test_data.config` though.

File: tests/config/test_data.config
Comment (nikhil) [reply tests/config/test_data.config]: Thanks, I fixed it

**Review by buehlere — APPROVED** [review]
Looks good!

**Comment by huyu335** [comment]
Could we please keep it the MR open for now? I'll need to check clinical needs for this module.

File: modules/msk/snppileup/meta.yml
Comment (anoronh4) [inline modules/msk/snppileup/meta.yml]: i think we no longer need `meta2` right? just `meta` and `meta1` now

**Review by anoronh4 — CHANGES_REQUESTED** [review]

File: modules/msk/snppileup/meta.yml
Comment (nikhil) [reply modules/msk/snppileup/meta.yml]: Sounds good, I removed it!

**Comment by nikhil** [comment]
> Could we please keep it the MR open for now? I'll need to check clinical needs for this module.

I am going to add you as a reviewer, let me know when you get the clinical needs and if its met with this module

**Review by anoronh4 — APPROVED** [review]

**Comment by huyu335** [comment]
@nikhil , thanks a lot for working on the module.
After checking clinical needs, we have one more bam file added to the command:

snp-pileup ${dbsnp} ${prefix}.snp_pileup.gz ${normal} ${tumor} **${unmatched-normal}**

This is usually for pool control bam. Can be optional if not available.
Do you want me to add it in, or you'd like to do from your end?

**Comment by rhshah** [comment]
@huyu335 i might be wrong but isn't this something that the executor will take care of providing the normal argument based on it being matched or unmatched

**Comment by huyu335** [comment]
> @huyu335 i might be wrong but isn't this something that the executor will take care of providing the normal argument based on it being matched or unmatched

From the clinical version of snp-pileup-wrapper.R:
parser$add_argument('-un', '--unmatched-normal-BAMS', required = FALSE, default = FALSE,
                    help = 'full path(s) as quoted string to unmatched normal BAM(s) to use for log ratio normalization, e.g. "<some/path>/*-NS_*bam"')
So we add BloodPoolNormal bam in the command in addition to 12345678-N and 12345678-T bam files.

I guess it's very unique to clinical side(?). If so, I am actually okay to just add something like ${other_sequence_file} at the end of the command, and we can add on the workflow side.

**Comment by rhshah** [comment]
Is this like a feature added to that wrapper on clinical side? As I dont see that option on the current main branch of facets-suite for that script

**Comment by huyu335** [comment]
> Is this like a feature added to that wrapper on clinical side? As I dont see that option on the current main branch of facets-suite for that script

Nope, not a feature nor a self-defined wrapper. This R script is from facets suite itself, clinical side just added one additional argument. Please note we don’t need to use this wrapper at all. I put the argument there since you were asking about matched / unmatched.
Basically you can consider clinical side has three bam files for each snp-pileup process, instead of two. Also, according to snp-pileup help, it should accept multiple bam files not just two.

for your reference:
https://github.com/mskcc/facets-suite/blob/master/snp-pileup-wrapper.R

**Comment by rhshah** [comment]
Make sense then should we extend the functionality to accommodate additional use cases:
- unmatched normal
- multiple BAM files?

**Comment by huyu335** [comment]
> Make sense then should we extend the functionality to accommodate additional use cases:
> 
> * unmatched normal
> * multiple BAM files?

Let’s do the second one, more leaning towards snp-pileup usage. My recommendation would be adding a variable silimar to ‘args’ from task, in case there will be use cases that need more than 3 bam files.

**Comment by nikhil** [comment]
Sounds good, I will do the second option to optionally add more bams after the tumor normal pair.

**Comment by nikhil** [comment]
okay changes have been added, this is ready for review

**Review by huyu335 — APPROVED** [review]

**Comment by huyu335** [comment]
Thanks a lot, @nikhil ! Looks great!

**Pull request merged**
